core/trino-main/src/main/java/io/trino/server/DynamicFilterService.java

```diff
@@ -672,24 +687,45 @@ public String toString()
         // This should not be a ConcurrentHashMap because we want to prevent concurrent addition of new consumers during the
         // removal of existing consumers from this map in addDynamicFilters. This ensures that new consumers don't miss filter completion.
         private final Map<DynamicFilterId, List<Consumer<Map<DynamicFilterId, Domain>>>> dynamicFilterConsumers = new HashMap<>();
-        private final long queryStartTime = System.nanoTime();
+        private final AtomicLong queryAttemptStartTime = new AtomicLong(System.nanoTime());
+        private final AtomicInteger currentAttemptId = new AtomicInteger(0);
 
         private DynamicFilterContext(
                 Session session,
                 Set<DynamicFilterId> dynamicFilters,
-                Map<DynamicFilterId, SettableFuture<Void>> lazyDynamicFilters,
+                Set<DynamicFilterId> lazyDynamicFilters,
                 Set<DynamicFilterId> replicatedDynamicFilters)
         {
             this.session = requireNonNull(session, "session is null");
             this.dynamicFilters = requireNonNull(dynamicFilters, "dynamicFilters is null");
-            this.lazyDynamicFilters = requireNonNull(lazyDynamicFilters, "lazyDynamicFilters is null");
+            requireNonNull(lazyDynamicFilters, "lazyDynamicFilters is null");
+            this.lazyDynamicFilters = new AtomicReference<>(lazyDynamicFilters.stream()
+                    .collect(toImmutableMap(identity(), filter -> SettableFuture.create())));
             this.replicatedDynamicFilters = requireNonNull(replicatedDynamicFilters, "replicatedDynamicFilters is null");
             dynamicFilters.forEach(filter -> {
                 taskDynamicFilters.put(filter, new ConcurrentHashMap<>());
                 dynamicFilterConsumers.put(filter, new ArrayList<>());
             });
         }
 
+        void resetForQueryRetry(int attemptId)
```

Comment:
Can we just create a new context (in `registerQueryRetry`) instead of reviving the old one? Seems easier and cleaner

Reply:
yup, updated to that now